services/warning/service.go

```diff
@@ -0,0 +1,87 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package warning
+
+import (
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd/pkg/deprecation"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/plugin/registry"
+	"github.com/containerd/containerd/plugins"
+	"github.com/containerd/containerd/services"
+)
+
+type Service interface {
+	EmitSchema1Warning()
```

Comment:
Simply the interface to `Emit(deprecation.Warning)`. Better to have a dependency in the interface than an interface that will keep adding new functions.

Reply:
Sure.  The difference here is that `Emit(deprecation.Warning)` will need to implement validation (and be either `Emit(deprecation.Warning) error`, `Emit(deprecation.Warning) bool`, or log a further warning in the daemon log) since a `deprecation.Warning` is a `string`.